reference_designs/ipu-xrt/matrix_multiplication_column/mm.cc

```diff
@@ -0,0 +1,216 @@
+          aie::store_v(pC2, C11.template to_vector<T_out>());
+          pC2 += MMUL::size_C;
+        }
+    }
+
+  event1();
+}
+
+template <unsigned m, unsigned k, unsigned n>
+void matmul_vectorized_4x4x4_i16_i16(const int16 *__restrict pA,
+                                     const int16 *__restrict pB,
+                                     int16 *__restrict pC) {
+  // matmul_vectorized operates on two 4x4 input blocks of A, and two 4x4 input
+  // blocks of B in each iteration. Make sure we have at least 2 blocks in each
+  // dimension, and that our input matrix is evenly divisible.
+  constexpr int r = 4;
+  constexpr int s = 4;
+  constexpr int t = 4;
+  static_assert(m % (2 * r) == 0 && m / (2 * r) > 0);
+  static_assert(k % (2 * s) == 0 && k / (2 * s) > 0);
+  static_assert(n % (2 * t) == 0 && n / (2 * t) > 0);
+  return matmul_vectorized<int16, int16, m / r, k / s, n / t, r, s, t>(pA, pB,
+                                                                       pC);
+}
+
+template <unsigned m, unsigned k, unsigned n>
+void matmul_vectorized_4x8x4_bf16_bf16(const bfloat16 *__restrict pA,
+                                       const bfloat16 *__restrict pB,
+                                       bfloat16 *__restrict pC) {
+  constexpr int r = 4;
+  constexpr int s = 8;
+  constexpr int t = 4;
+  static_assert(m % (2 * r) == 0 && m / (2 * r) > 0);
+  static_assert(k % (2 * s) == 0 && k / (2 * s) > 0);
+  static_assert(n % (2 * t) == 0 && n / (2 * t) > 0);
+  return matmul_vectorized<bfloat16, bfloat16, m / r, k / s, n / t, r, s, t>(
+      pA, pB, pC);
+}
+
+template <unsigned m, unsigned k, unsigned n>
+void matmul_vectorized_4x8x4_bf16_f32(const bfloat16 *__restrict pA,
+                                      const bfloat16 *__restrict pB,
+                                      float *__restrict pC) {
+  constexpr int r = 4;
+  constexpr int s = 8;
+  constexpr int t = 4;
+  static_assert(m % (2 * r) == 0 && m / (2 * r) > 0);
+  static_assert(k % (2 * s) == 0 && k / (2 * s) > 0);
+  static_assert(n % (2 * t) == 0 && n / (2 * t) > 0);
+  return matmul_vectorized<bfloat16, float, m / r, k / s, n / t, r, s, t>(
+      pA, pB, pC);
+}
+
+extern "C" {
+
+#define combos(X)                                                              \
+  X(int16, i16, int16, i16, 4, 4, 4)                                           \
+  X(bfloat16, bf16, bfloat16, bf16, 4, 8, 4)                                   \
+  X(bfloat16, bf16, float, f32, 4, 8, 4)
+
+#define matmul_vectorized_c_func(ctype_in, mlir_type_in, ctype_out,            \
+                                 mlir_type_out, r, s, t)                       \
+  void matmul_##mlir_type_in##_##mlir_type_out(ctype_in *a_in, ctype_in *b_in, \
+                                               ctype_out *c_out) {             \
+    matmul_vectorized_##r##x##s##x##t##_##mlir_type_in##_##mlir_type_out<      \
+        64, 32, 64>(a_in, b_in, c_out);                                        \
```

Comment:
I know we probably don't want to use templating any more, and we don't have any way to generate the C code from Python easily, but are there any alternatives to hard-coding the 64x32x64 size here? Maybe, we could at least make it an argument to the `combos()` macro, so it would be easy to generate a bunch of different tile sizes by adding lines to the same macro?

Something like:
```
#define combos(X)                                                              \
  X(int16, i16, int16, i16, 4, 4, 4, 64, 32, 64)                                           \
  X(bfloat16, bf16, bfloat16, bf16, 4, 8, 4, 64, 32, 64)                                   \
  X(bfloat16, bf16, float, f32, 4, 8, 4, 64, 32, 64)
```

(Then you could generate all necessary functions for another tile size by just adding a line below that with differing last three arguments.)

and then here (added M, N, K to function):
```
#define matmul_vectorized_c_func(ctype_in, mlir_type_in, ctype_out,            \
                                 mlir_type_out, r, s, t, M, N, K)                       \
  void matmul_##mlir_type_in##_##mlir_type_out(ctype_in *a_in, ctype_in *b_in, \
                                               ctype_out *c_out) {             \
    matmul_vectorized_##r##x##s##x##t##_##mlir_type_in##_##mlir_type_out<      \
        M, N, K>(a_in, b_in, c_out);  
```

and analogously for the scalar and zeroing functions below.

Just a thought ... hard-coding the size here just seems like a bit of a regression from how easy it was to experiment with different sizes before.